datafusion/sql/src/unparser/dialect.rs

```diff
@@ -171,6 +185,67 @@ impl Dialect for PostgreSqlDialect {
     fn float64_ast_dtype(&self) -> sqlparser::ast::DataType {
         sqlparser::ast::DataType::DoublePrecision
     }
+
+    fn scalar_function_to_sql_overrides(
+        &self,
+        unparser: &Unparser,
+        func_name: &str,
+        args: &[Expr],
+    ) -> Result<Option<ast::Expr>> {
+        if func_name == "round" {
+            return Ok(Some(
+                self.round_to_sql_enforce_numeric(unparser, func_name, args)?,
+            ));
+        }
+
+        Ok(None)
+    }
+}
+
+impl PostgreSqlDialect {
+    fn round_to_sql_enforce_numeric(
+        &self,
+        unparser: &Unparser,
+        func_name: &str,
+        args: &[Expr],
+    ) -> Result<ast::Expr> {
+        let mut args = unparser.function_args_to_sql(args)?;
+
+        // Enforce the first argument to be Numeric
+        if let Some(ast::FunctionArg::Unnamed(ast::FunctionArgExpr::Expr(expr))) =
+            args.first_mut()
+        {
+            if let ast::Expr::Cast { data_type, .. } = expr {
```

Comment:
Rather than checking after unparsing, maybe this code could check *before* unparsing  -- as in get the type of Expr in `args[0]` and use that to decide if a cast is needed

Reply:
That is a good idea 👍. Unfortunately, we can't enforce `NUMERIC` before unparsing, as we operate on the `Arrow` DataType level, which includes `FLOAT, DOUBLE`, etc., and these types will be converted to `FLOAT, DECIMAL, or DOUBLE PRECISION` types and we can't enforce NUMERIC before unparsing. Since we require to do unparsing anyway, I think it is better to have all the relevant logic in the same place—after unparsing.